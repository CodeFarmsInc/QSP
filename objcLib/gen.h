#ifndef ZZ_CODEGEN_INCLUDE
#define ZZ_CODEGEN_INCLUDE

#include "lib/general.h"

class Generator;
class Organization;
class Participant;
class OrgType;
class PartType;
class Iterator;
class FixedType;
class ClassLink;
class ApplClass;

#include "tmp/orgs.h"
#include "tmp/orgs1.h"
#include "tmp/parts.h"
#include "tmp/parts1.h"
#include "tmp/parts2.h"
#include "tmp/orgName.h"
#include "tmp/partName.h"
#include "tmp/types.h"
#include "tmp/types1.h"
#include "tmp/partTypes.h"
#include "tmp/partTypes1.h"
#include "tmp/partTypes2.h"
#include "tmp/inherits.h"
#include "tmp/iters.h"
#include "tmp/iters1.h"
#include "tmp/orgTypeName.h"
#include "tmp/fileName.h"
#include "tmp/partTypeName.h"
#include "tmp/iterName.h"
#include "tmp/inhName.h"
#include "tmp/fixedName.h"
#include "tmp/inhMatch.h"
#include "tmp/inhFixed.h"
#include "tmp/fixedTypes.h"
#include "tmp/fixedTypes1.h"
#include "tmp/myType.h"
#include "tmp/myPartType.h"
#include "tmp/toBase.h"
#include "tmp/classes.h"
#include "tmp/classes1.h"
#include "tmp/dependsOn.h"
#include "tmp/dependsOn1.h"
#include "tmp/usedBy.h"
#include "tmp/usedBy1.h"
#include "tmp/usedBy2.h"
#include "tmp/cName.h"

// -----------------------------------------------
typedef class orgs_LinkedList2 orgs;
typedef class parts_Aggregate2 parts;
typedef class orgName_Name orgName;
typedef class partName_Name partName;
typedef class types_LinkedList2 types;
typedef class partTypes_Aggregate2 partTypes;
typedef class inherits_SingleLink inherits;
typedef class iters_LinkedList1 iters;
typedef class orgTypeName_Name orgTypeName;
typedef class fileName_Name fileName;
typedef class partTypeName_Name partTypeName;
typedef class iterName_Name iterName;
typedef class inhName_Name inhName;
typedef class fixedName_Name fixedName;
typedef class inhMatch_SingleLink inhMatch;
typedef class inhFixed_SingleLink inhFixed;
typedef class fixedTypes_LinkedList1 fixedTypes;
typedef class myType_SingleLink myType;
typedef class myPartType_SingleLink myPartType;
typedef class toBase_SingleLink toBase;
typedef class classes_Hash classes;
typedef class dependsOn_LinkedList1 dependsOn;
typedef class usedBy_Aggregate2 usedBy;
typedef class cName_Name cName;
// -----------------------------------------------


class ZZ_Generator {
public:
    orgs_LinkedList2Parent ZZorgs;
    types_LinkedList2Parent ZZtypes;
    classes_HashHolder ZZclasses;
};

class ZZ_Organization {
public:
    orgs_LinkedList2Child ZZorgs;
    parts_Aggregate2Parent ZZparts;
    orgName_NameParent ZZorgName;
    myType_SingleLinkParent ZZmyType;
};

class ZZ_Participant {
public:
    parts_Aggregate2Child ZZparts;
    partName_NameParent ZZpartName;
    myPartType_SingleLinkParent ZZmyPartType;
    usedBy_Aggregate2Child ZZusedBy;
};

class ZZ_OrgType {
public:
    types_LinkedList2Child ZZtypes;
    partTypes_Aggregate2Parent ZZpartTypes;
    inherits_SingleLinkParent ZZinherits;
    iters_LinkedList1Parent ZZiters;
    orgTypeName_NameParent ZZorgTypeName;
    fileName_NameParent ZZfileName;
    fixedTypes_LinkedList1Parent ZZfixedTypes;
};

class ZZ_PartType {
public:
    partTypes_Aggregate2Child ZZpartTypes;
    partTypeName_NameParent ZZpartTypeName;
    inhName_NameParent ZZinhName;
    inhMatch_SingleLinkParent ZZinhMatch;
};

class ZZ_Iterator {
public:
    iters_LinkedList1Child ZZiters;
    iterName_NameParent ZZiterName;
};

class ZZ_FixedType {
public:
    fixedName_NameParent ZZfixedName;
    inhFixed_SingleLinkParent ZZinhFixed;
    fixedTypes_LinkedList1Child ZZfixedTypes;
};

class ZZ_ClassLink {
public:
    toBase_SingleLinkParent ZZtoBase;
    dependsOn_LinkedList1Child ZZdependsOn;
};

class ZZ_ApplClass {
public:
    classes_HashElement ZZclasses;
    dependsOn_LinkedList1Parent ZZdependsOn;
    usedBy_Aggregate2Parent ZZusedBy;
    cName_NameParent ZZcName;
};

typedef orgs_LinkedList2Iterator orgs_Iterator;
typedef parts_Aggregate2Iterator parts_Iterator;
typedef types_LinkedList2Iterator types_Iterator;
typedef partTypes_Aggregate2Iterator partTypes_Iterator;
typedef iters_LinkedList1Iterator iters_Iterator;
typedef fixedTypes_LinkedList1Iterator fixedTypes_Iterator;
typedef classes_HashIterator classes_Iterator;
typedef dependsOn_LinkedList1Iterator dependsOn_Iterator;
typedef usedBy_Aggregate2Iterator usedBy_Iterator;

#endif // ZZ_CODEGEN_INCLUDE
