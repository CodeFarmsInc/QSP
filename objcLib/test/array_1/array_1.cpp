// ------------------------------------------------------------------------
// This is test16c from the DOL regression suite, and the results
// should be identical except for the type information which is formatted 
// differently.
// This program tests not only Array, but also Ring1, SingleLink and
// Name organizations.
//                                        Jiri Soukup, Mar.10, 2004
// ------------------------------------------------------------------------
// The input file has two types of records:
// P name age salary              // person description
// C name1 name2 numberOfChidren  // couple description
// ------------------------------------------------------------------------

#include <stdio.h>
#include <string.h>
#include "gen.h"

class PrtObject {
public:
    virtual void prt(FILE *fp){fprintf(fp,"PrtObject\n");}
};

class Header : public PrtObject {
public:
    ZZ_Header ZZds;
    void prt(FILE *fp){fprintf(fp,"Header\n");}
};

class Person : public PrtObject {
public:
    ZZ_Person ZZds;
    int age;
    float salary;
    void prt(FILE *fp){fprintf(fp,"Person=%s age=%d salary=%6.2f\n",
                                                   name.get(this),age,salary);}
};

class Couple : public PrtObject {
public:
    ZZ_Couple ZZds;
    int numChild;
    void prt(FILE *fp){fprintf(fp,"Couple numChild=%d/n",numChild);}
};

#define BSIZE 80

int main(int argc,char **argv) {
    char bf[BSIZE],name1[BSIZE],name2[BSIZE];
    Person *getPerson(Person *p,char *nm);
    void printObj(char *obj,int typeInd,int size,const char *priv);
    void printAll(FILE *fp,Header *h);
    FILE *file1,*file2;
    Header *h;
    char c,*n1,*v[1],*t[1];
    int i,n,age,child,*ci;
    float salary;
    Person *sp,*pp,*p1,*p2;
    Couple *cp;

    if(argc<=1)file1=stdin;  else file1=fopen(argv[1],"r");
    if(argc<=2)file2=stdout; else file2=fopen(argv[2],"w");

    h=new Header;
    sp=(Person *)NULL;
    n=i=0;
    while(fgets(bf,BSIZE,file1)){
        sscanf(bf,"%c",&c);
        if(c=='P'){
            if(n<0){printf("all P records must preceed C records\n");return(0);}
            n++;
            sscanf(bf,"%c %s %d %f",&c,name1,&age,&salary);
            pp=new Person;
            pp->age=age;
            pp->salary=salary;
            n1=new char[strlen(name1)+1]; strcpy(n1,name1);
            name.add(pp,n1);
            sp=people.add(sp,pp);
        }
        else if(c=='C'){
            if(n>0){
                (void)arr.form(h,n/2,3);
                if(!cArr.formed(h))ci=cArr.form(h,n/2,3); // test of formed()
                if(!cArr.formed(h))ci=cArr.form(h,n/2,3); // should not form
                n= -1;
            }
            sscanf(bf,"%c %s %s %d",&c,name1,name2,&child);
            p1=getPerson(sp,name1);
            p2=getPerson(sp,name2);
            if(p1 && p2){
                cp=arr.ind(h,i);
                per1.add(cp,p1);
                per2.add(cp,p2);
                cp->numChild=child;
                ci[i]=child;
                i++;
            }
            else printf("wrong couple definition %s %s\n",name1,name2);
        }
    }
    fclose(file1);

    printAll(file2,h);
    return(0);
}
/* ------------------------------------------------------------------ */
Person *getPerson(Person *sp,char *nm)
{
    Person *pp;
    char *nn;
    people_iterator pIter;

    pIter.start(sp);
    for(pp= ++pIter;pp;pp= ++pIter){ // See NOTE at the end of the code
        nn=name.get(pp);
        if(!strcmp(nm,nn))return(pp);
    }
    return((Person *)NULL);
}
/* ------------------------------------------------------------------ */
void printAll(FILE *file2,Header *h)
{
    int i,sz;
    int *ci;
    char *n1,*n2;
    Couple *ca;
    Person *p1,*p2,*pp;
    people_iterator pIt;

    sz=arr.size(h);
    ca=arr.ind(h,0);
    ci=cArr.ind(h,0);

    for(i=0;i<sz;i++,ca++){
        p1=per1.target(ca);
        n1=name.get(p1);
        p2=per2.target(ca);
        n2=name.get(p2);

        if(i==0 && p1){ // before the first couple, print the list of all people
            pIt.start(p1);
            ITERATE(pIt,pp){
                pp->prt(file2);
            }
            fprintf(file2,"\n");
        }

        // print the next couple
        fprintf(file2,
          "%s(%d,%6.2f) %s(%d,%6.2f) child=%d\n",
               n1,p1->age,p1->salary,n2,p2->age,p2->salary,ca->numChild);
        if(ca->numChild!=ci[i])fprintf(file2,
                    "   disagreement: i=%d ca=%d ci=%d\n",i,ca->numChild,ci[i]);
    }
}
#include "gen.cpp"

// NOTE: The most convenient way of iterating over a set is to use a while()
//       loop. In this situation, you can simply use:
//                while(pp=pIter++){
//       Unfortunately, many compilers issue warnings for this code.
//       Using the syntactically correct for() is unwieldy:
//                for(pp= ++pIter;pp;pp= ++pIter){
//       but can be hidden under a macro which is provided by the library:
//                ITERATE(pIter,pp){
