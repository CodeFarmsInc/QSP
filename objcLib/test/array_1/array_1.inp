P Adam 22 12.5
P Eva  19 9.3
P Michael 65 400.0
P Gizella 62 91.1
P Grace 73 11.1
P Peter 53 12.5
P Alena 45 120.3
P Tom 52   55.35
C Adam Eva 10
C Michael Gizella 2
C Peter Alena 0
