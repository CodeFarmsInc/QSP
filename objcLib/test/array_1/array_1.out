Person=Tom age=52 salary= 55.35
Person=Alena age=45 salary=120.30
Person=Peter age=53 salary= 12.50
Person=Grace age=73 salary= 11.10
Person=Gizella age=62 salary= 91.10
Person=Michael age=65 salary=400.00
Person=Eva age=19 salary=  9.30
Person=Adam age=22 salary= 12.50

Adam(22, 12.50) Eva(19,  9.30) child=10
Michael(65,400.00) Gizella(62, 91.10) child=2
Peter(53, 12.50) Alena(45,120.30) child=0
